pennylane/ops/qubit/state_preparation.py

```diff
@@ -235,6 +259,73 @@ def state_vector(self, wire_order: Optional[WiresLike] = None):
             transpose_axes = [0] + [a + 1 for a in transpose_axes]
         return math.transpose(op_vector, transpose_axes)
 
+    @staticmethod
+    def _preprocess(state, wires, pad_with, normalize):
+        """Validate and pre-process inputs as follows:
+
+        * If state is batched, the processing that follows is applied to each state set in the batch.
+        * Check that the state tensor is one-dimensional.
+        * If pad_with is None, check that the last dimension of the state tensor
+          has length :math:`2^n` where :math:`n` is the number of qubits. Else check that the
+          last dimension of the state tensor is not larger than :math:`2^n` and pad state
+          with value if necessary.
+        * If normalize is false, check that last dimension of state is normalised to one. Else, normalise the
+          state tensor.
+        """
+        if isinstance(state, (list, tuple)):
+            state = math.array(state)
+
+        shape = math.shape(state)
+
+        # check shape
+        if len(shape) not in (1, 2):
+            raise ValueError(
+                f"State must be a one-dimensional tensor, or two-dimensional with batching; got shape {shape}."
+            )
+
+        n_states = shape[-1]
+        dim = 2 ** len(Wires(wires))
+        if pad_with is None and n_states != dim:
+            raise ValueError(
+                f"State must be of length {dim}; got length {n_states}. "
+                f"Use the 'pad_with' argument for automated padding."
+            )
+
+        if pad_with is not None:
+            if n_states > dim:
+                raise ValueError(
+                    f"Input state must be of length {dim} or "
+                    f"smaller to be padded; got length {n_states}."
+                )
+
+            # pad
+            if n_states < dim:
+                padding = [pad_with] * (dim - n_states)
+                if len(shape) > 1:
+                    padding = [padding] * shape[0]
+                padding = math.convert_like(padding, state)
+                state = math.hstack([state, padding])
+
+        # normalize
+        if "int" in str(state.dtype):
+            state = math.cast_like(state, 0.0)
+        norm = math.linalg.norm(state, axis=-1)
```

Comment:
I think computing `norm` is the computation bottleneck and not renormalization via division. This makes me wonder, how does `normalization` keyword actually helps lol.

Reply:
I see your point, I can compute the norm only if `normalize = True`.
Should I write `math.is_abstract(state)` then?

Reply:
That would be ideal. Yes, I think you can do the renormalization when the `state` is `abstract` and when `normalize=True`.

Reply:
Oh but now that I see it, then you would be suggesting to remove the error message that an input is not normalized

Reply:
Please add a `TODO` for this as well.